### DISCUSSION TITLE
Bricked Dremel 3D20 - No USB. Can I install Marlin using ST-Link?

### DISCUSSION THREAD

**Discussion by terenceang** [moonglow/FlashForge_Marlin#101] [General]
My Dremel 3d20 was bricked during a firmware update (I wasn't around, a friend did the upgrade, not sure what happened)
My computer won't recognise the board, no USB, only Lights on the board is 24v, 3.3v and USB.

I am trying to revive it by flashing it with Marlin using ST-Link, but it didn't work. ST-<Link can see the board, erase and flash it, but still not working. 
Can anyone point me in the right direction on how to reflash it?

Thanks in advance. 

Note. I do have another working 3D20.

**Comment by moonglow** [comment]
Hi @terenceang !

>I am trying to revive it by flashing it with Marlin using ST-Link, but it didn't work. 

You need to write **unencrypted** firmware, can you reproduce step-by-step how did you try it ? You need also setup correct flash start address. If you already erase your main flash of MCU and lost your bootloader and so on, you need to do next thing:

Full recovery steps:

1. Write new bootloader binary ( it from Dreamer, but it also works ) you need to write at start of flash `0x08000000`  [dreamer_bootloader_v1.4_20161121_no_check.zip](https://github.com/moonglow/FlashForge_Marlin/files/8244294/dreamer_bootloader_v1.4_20161121_no_check.zip) ( bootloader without MCU ID check )
2. Write unencrypted firmware to it location `0x08010000`  [app_firmware_dremel.zip](https://github.com/moonglow/FlashForge_Marlin/files/8244403/app_firmware_dremel.zip)


You you did not lost your original bootloader just use step 2. 

p.s.: if your bootloader is still present please make full flash dump and share it here it will helpful for other persons who want to recovery their printers too.

**Reply by terenceang** [reply]
Hi, 

Thanks a mil, for the reply. 

I followed your step in got Marlin to work on it. 
I am now trying to compile and mod for the MKS version

The original bootloader is lost, I will open up the other printer and make a backup of it and post it here. 

Once again, Thanks...

**Reply by Thatindianbruh** [reply]
Any chance you have that bootloader? I may have bricked mine. Flashing without securing first.

**Reply by moonglow** [reply]
@Thatindianbruh

1. You can try bootloader posted  here: https://github.com/moonglow/FlashForge_Marlin/discussions/101#discussioncomment-2356978 if your erase your original bootloader
2. If you just write invalid date to firmware location, you can reflash it with actual firmware ( unencrypted ) and all be OK

**Reply by Thatindianbruh** [reply]
Thank you. I'll give that a shot Tuesday when my ST link adapter gets delivered. 

how do i confirm the bootloader is ok? 
It boots using the sd card. I can see the dremel 3d idea builder and the orange logo, but after that nothing. I can change the images and they will show also at boot.

**Reply by tracedgod** [reply]
> Thank you. I'll give that a shot Tuesday when my ST link adapter gets delivered.
> 
> how do i confirm the bootloader is ok? It boots using the sd card. I can see the dremel 3d idea builder and the orange logo, but after that nothing. I can change the images and they will show also at boot.

If you can see the boot logo when the printer turns on, there is a good chance the bootloader is OK. The problem sounds like it's more the section of flash where your original firmware was is corrupt. When you get the ST Link, try flashing the original firmware for your printer first and see if it boots back up. If you still get nothing, back-up your chip flash (assuming you are using [STM32CubeProgrammer](https://www.st.com/en/development-tools/stm32cubeprog.html) or [STM32 ST-LINK/V2](https://www.st.com/en/development-tools/st-link-v2.html)) and do a full chip erase, then flash the bootloader binary provided above then the original firmware and that should get you up and running.

On another note, there might be a chance you can get the bootloader to flash your firmware as is with no ST Link adapter, though I'm not familiar with the differences with the Dremel bootloader and FlashForge bootloader, but typically the bootloader will check the SD card for a file on the internal SD at ``sys/dreamer.bin`` and begin writing firmware if it is found, perhaps you could try this with the original firmware and see if it does anything for you? I'm not sure what the bootloader could be checking for in this case, so maybe try the ``dreamer.bin`` filename first, if you get nothing, maybe try ``3d20.bin`` or ``dremel.bin``?

**Reply by Thatindianbruh** [reply]
Thank you for your insight once again. It helped tremendously. 

Turns out my bootloader was fine and I was able to flash the most current version of [colorui](https://github.com/moonglow/FlashForge_Marlin/releases/download/v0.15.0/color_ui_10022022.zip) from the releases tab at `0x08010000`. 

I wiped sector `0x08010000` first and then flashed [colorui](https://github.com/moonglow/FlashForge_Marlin/releases/download/v0.15.0/color_ui_10022022.zip) 

I did attempt flashing the firmware using the sdcard method by leaving all variations of the name in the sys folder and that wasn't helpful. I assume it had something to do with the bad firmware flash i did prior. 

I did find out that 3d20 requires the file to be called `dremel.bin`.

**Reply by moonglow** [reply]
Is all ok now ?

**Reply by Thatindianbruh** [reply]
yes! all ok now. Thank you

**Reply by 1Techpro** [reply]
Hi, I  was able to get my printer running again after failed update, but now I have no logo Dremel screen when booting, it is white for a couple seconds and then Marlin logo screen. Is this normal?  Also thanks for the great work!

**Reply by tracedgod** [reply]
> Hi, I was able to get my printer running again after failed update, but now I have no logo Dremel screen when booting, it is white for a couple seconds and then Marlin logo screen. Is this normal? Also thanks for the great work!

This would happen if the bootloader does not find the bitmap file located on the internal SD card, do you have the original files still located on the card?

**Comment by GrandWaz00** [comment]
Here you go
[dremel_TFcard - no firmware.zip](https://github.com/moonglow/FlashForge_Marlin/files/10528679/dremel_TFcard.-.no.firmware.zip)

**Reply by 1Techpro** [reply]
I have all the files as well as the latest firmware on the internal SD card, I had to erase the entire memory when I was stuck/bricked with only a black screen. so I was wondering if somebody had a copy from address 0x10000000. I used the above bootloader and have to use the above app.bin to get it running. Now it seems like it's not registering the internal SD card at all. I am stuck at version 2.09.3.

**Reply by GrandWaz00** [reply]
I have a bootloader (contents of flash from 0x08000000-0x0800FFFF), but I don't know how to find or zap the board serial# (or apply the "board hash patch"), therefore (according to previous writings by moonglow) it would be unusable on anyone else's machine.

@moonglow please advise.

**Reply by 1Techpro** [reply]
I think my problem is now my require an unencrypted firmware, and I'm not entirely sure how to decrypt. Is it possible to post a version of the decrypted firmware? Or is this not the problem.

**Reply by moonglow** [reply]
@GrandWaz00 
Bootloader board hash location is: `0x0800C000`, size `8` bytes, i think i need to add algorithm to calculate valid hash for any board if someone do not want to use patched bootloader.

@1Techpro 
You can decrypt any FF firmware ( original or my marlin based ) with my firmware decryption/encryption tool:  [LINK](https://github.com/moonglow/flashforge_firmware_tool). If you want to write firmware with JTAG you must use decrypted firmware, because MCU can only process know instruction set and not encrypted data

**Reply by 1Techpro** [reply]
I'm have trouble using the firmware decryption/encryption tool, is there anymore information on how to use? When I try to run it open's and closes. I have no experience with Pelles C. When I run the ff_tool project I get input/output path dose not exist. Not sure what I'm doing wrong. Thanks again for the help!

**Reply by 1Techpro** [reply]
Or can you add unencrypted Dremel files to your releases?

**Reply by moonglow** [reply]
@1Techpro 

>, is there anymore information on how to use?

Yep you can find it on project page.  Under `Usage` headline.

>When I try to run it open's and closes. 

This is command line utility, you need to run it from console window with specific arguments what described on project page

>Or can you add unencrypted Dremel files to your releases?

Unencrypted files will not supported by Dremel/FF bootloader and will break printer, so this is not a good idea to put such files to Releases, this is useful only for manual flashing or other not regular needs, but if you have a problem to decrypt firmware, i attach it for you here:

# DO NOT USE THIS FILES WITH BOOTLOADER 
[decrypted_dremel.zip](https://github.com/moonglow/FlashForge_Marlin/files/10578812/decrypted_dremel.zip)

**Reply by 1Techpro** [reply]
Thank you very much, I will do some learning on how to use your decryptor, but I greatly appreciate you uploading this file so I can try the new release.

**Reply by GrandWaz00** [reply]
I put a Windows binary here: https://github.com/moonglow/FlashForge_Marlin/discussions/139
